---
title: "A Thiopurine-like Mutagenic Process Defines TGCT Subtypes"
author: "Tongwu Zhang"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

## Overview
This R Markdown document reproduces the main analyses and figures for TCGA-TGCT WGS dataset

Manuscript title: A Thiopurine-like Mutagenic Process Defines TGCT Subtype

Fig. 1: Distinct Genomic Landscapes of Seminomas and Non-Seminomas in Testicular Germ Cell Tumors (TGCTs). 
---

## Load Required Libraries and Set Plotting Styles 
(You may need to install some packages if missing)

```{r}

source('../functions/Sherlock_functions.R')

set_wd()
libztw()
pdfhr()
pdfhr2()

```

## Fig. 1a.
Comparison of key genomic features between seminomas (blue) and non-seminomas (purple), including tumor mutational burden (TMB; mutations/Mb), percentage genome altered (PGA; WGD-adjusted), number of structural variants (SVs), and number of transposable element insertions (TEs). P-values were obtained from linear regression models adjusted for age and tumor purity. For TMB, tumor ploidy was also included as a covariate. Multiple testing was corrected using the Benjamini–Hochberg FDR method.

```{r,fig.width=12}
# Load R object WGS analysis results --------------------------------------

load('../data/wgs_data.RData')
load('../data/wgs_clinical.RData')
load('../data/ngspurity_data.RData')

tgct_variable <- NULL
# Numbers -----------------------------------------------------------------

# TMB
grch38 <- read_delim(
  '../data/Homo_sapiens_assembly38.fasta.fai',
  delim = '\t',
  col_names = FALSE
) #%>% slice(1:24)
grch38_size <- grch38 %>%
  filter(X1 %in% paste0('chr', c(1:22, 'X', 'Y', 'M'))) %>%
  summarise(total = sum(X2)) %>%
  pull(total)
mafdata <- read_delim('../data/all_mutianno.txt.gz', delim = '\t', col_names = T) %>%
  clean_names()
tgct_tmb <- mafdata %>%
  filter(chr %in% paste0('chr', c(1:22, 'X', 'Y', 'M'))) %>%
  count(tumor_sample_barcode) %>%
  select(Tumor_Barcode = tumor_sample_barcode, Mutations = n) %>%
  mutate(TMB = 10^6 * Mutations / grch38_size)

tgct_tmb %>% pivot_longer(-Tumor_Barcode)


# PGA
load('../data/tgct_PGA.RData', verbose = T)
tgct_PGA %>% select(Tumor_Barcode, PGA_WGD)

# SVs
load('../data/SV_Fusion.RData', verbose = T)

tgct_sv <- wgs_data %>%
  select(Subject, Tumor_Barcode) %>%
  left_join(svdata %>% count(Tumor_Barcode = SAMPLE, name = 'SVs')) %>%
  mutate(SVs = replace_na(SVs, 0))

# TE insertions
load('../data/trafic.RData', verbose = T)


tgct_variable <- tgct_tmb %>%
  left_join(tgct_PGA) %>%
  left_join(tgct_sv) %>%
  left_join(
    trafic_count %>% select(Tumor_Barcode, Total_TE:Total_L1_SOMATIC)
  ) %>%
  left_join(
    ngspurity %>% select(Tumor_Barcode, BB_Purity)
  ) %>%
  left_join(wgs_clinical %>% select(Subject, age = age_at_diagnosis)) %>%
  select(-Subject) %>%
  pivot_longer(-Tumor_Barcode)


#save(tgct_variable, file = 'tgct_variabe.RData')


# comparision between subtypes and genomic alterations

wgs_data %>%
  left_join(tgct_variable) %>%
  group_by(name) %>%
  do(tidy(lm(value ~ Subtype,data=.)))


pdata <- wgs_data %>%
  left_join(
    tgct_variable %>%
      filter(name %in% c('TMB', 'PGA_WGD', 'SVs', 'Total_TE'))
  ) %>%
  left_join(ngspurity %>% select(Tumor_Barcode, BB_Purity,BB_Ploidy)) %>%
  left_join(
    wgs_clinical %>% select(Subject, age = age_at_diagnosis)
  )


pdata <- pdata %>%
  mutate(
    name = factor(
      name,
      levels = c('TMB', 'PGA_WGD', 'SVs', 'Total_TE'),
      labels = c(
        'TMB\n(mutations/Mb)',
        'PGA\n(WGD adjusted; %)',
        'SV\n(No. of events)',
        'TE\n(No. of events)'
      )
    )
  )

result <- pdata %>%
  group_by(name) %>%
  do(tidy(lm((value) ~ Subtype + age + BB_Purity   , data = .))) %>% #BB_Ploidy
  filter(term == 'SubtypeNon-Seminoma') %>%
  ungroup() %>%
  mutate(FDR = p.value)
#mutate(FDR = p.adjust(p.value, method = 'BH'))

result <- result %>%
  mutate(
    myformatted.p = if_else(
      FDR < 0.001,
      sprintf("P = %.2e", FDR),
      sprintf("P = %.3f", FDR)
    )
  ) %>%
  mutate(Subtype = "Non-Seminoma") %>%
  mutate(col = if_else(FDR < 0.05, ncicolpal[1], 'black')) %>%
  mutate(group1 = 'Seminoma', group2 = 'Non-Seminoma') %>%
  left_join(
    pdata %>% group_by(name) %>% summarise(y.position = 1.1 * max(value))
  )


# my_comparisons <-  list(c('Seminoma','Non-Seminoma'))
# stat.test <- pdata %>%
#   lm(value ~ Subtype + age + BB_Purity, comparisons = my_comparisons) %>%
#   rstatix::add_xy_position(x = "Subtype") %>%
#   mutate(myformatted.p = if_else(p<0.001,sprintf("P = %.2e",p),sprintf("P = %.2f",p))) %>%
#   mutate(Subtype = my_comparisons[[1]][1]) %>%
#   mutate(col=if_else(p<0.05,ncicolpal[1],'black'))

pdata %>%
  ggplot(aes(Subtype, value, fill = Subtype)) +
  geom_violin(trim = T, size = 0.1) +
  geom_boxplot(
    width = 0.3,
    fill = NA,
    color = 'black',
    outlier.shape = 21,
    outlier.size = 0.8
  ) +
  #ggbeeswarm::geom_quasirandom(pch=21,size=2,width = 0.2,color="white",stroke=0.2)+
  facet_wrap(~name, scale = 'free', nrow = 1) +
  labs(x = NULL, y = 'Genomic alteration value') +
  scale_fill_manual(values = subtypecol) +
  scale_y_continuous(breaks = pretty_breaks(n = 7)) +
  theme_ipsum_rc(
    axis_text_size = 15,
    axis_title_just = 'm',
    axis_title_size = 16,
    grid = 'XY',
    ticks = T
  ) +
  panel_border(color = 'black') +
  theme(
    axis.text.x = element_blank(),
    panel.spacing.x = unit(0.2, 'cm'),
    axis.ticks.x = element_blank(),
    plot.margin = margin(4, 4, 4, 4),
    legend.position = 'right',
    legend.text = element_text(size = 15),
    legend.title = element_text(size = 15),
    strip.text.x = element_text(hjust = 0.5, face = 'bold', size = 15)
  ) +
  # stat_pvalue_manual(
  #   data = result,
  #   col = 'col',
  #   label = "myformatted.p",
  #   size = 5
  # ) +
  scale_color_manual(values = c(ncicolpal[1], 'black')) +
  guides(color = 'none')

```
## Fig. 1b
Significantly mutated genes identified by the IntOGen pipeline across all TCGT samples. Dot size reflects combined -log10 q-value based on different driver gene algorithms. Colors denote functional annotation (Act: activating; LoF: loss-of-function).

```{r}

load('../data/intogene_drivers.RData')

## visulaizaiton

intogene %>% 
  mutate(SYMBOL = fct_reorder(SYMBOL,SAMPLES)) %>% 
  mutate(Frequency=SAMPLES/252) %>% 
  ggplot(aes(SYMBOL,Frequency,size=SAMPLES,fill=-log10(QVALUE_COMBINATION)))+
  geom_point(pch=21)+
  scale_y_continuous(breaks = pretty_breaks(),labels = percent_format())+
  scale_fill_viridis_c()+
  scale_size_binned()+
  labs(x=NULL,y='Mutated Frequency',fill="-log10(qvalue)\n",size='nSample')+
  theme_ipsum_rc(base_size = 14,axis_title_just = 'm',axis_title_size = 16,grid = 'XY',ticks = T)+
  theme(axis.text.x = element_text(angle = 90,hjust = 1,vjust = 0.5,face =3),
        legend.position = 'bottom',
        legend.key.height = unit(0.3,units = 'cm'),
        legend.key.width = unit(0.7,units = 'cm'),
        plot.margin = margin(4,4,4,4))+
  coord_cartesian(clip = 'off')+
  panel_border(color = 'black',size = 0.4)



intogene %>% 
  mutate(SYMBOL = fct_reorder(SYMBOL,SAMPLES)) %>% 
  mutate(Frequency=SAMPLES/252) %>% 
  ggplot(aes(SYMBOL,Frequency,size=-log10(QVALUE_COMBINATION),fill=ROLE))+
  geom_point(pch=21)+
  scale_y_continuous(breaks = pretty_breaks(),labels = percent_format())+
  scale_fill_bmj()+
  scale_size_binned()+
  labs(x=NULL,y='Mutated Frequency',size="-log10(qvalue)\n",fill='Role')+
  theme_ipsum_rc(base_size = 14,axis_title_just = 'm',axis_title_size = 16,grid = 'XY',ticks = T)+
  guides(fill=guide_legend(override.aes = list(size=4)))+
  theme(axis.text.x = element_text(angle = 90,hjust = 1,vjust = 0.5,face =3),
        legend.position = 'bottom',
        legend.key.height = unit(0.3,units = 'cm'),
        legend.key.width = unit(0.7,units = 'cm'),
        plot.margin = margin(4,4,4,4),
        legend.box.margin = margin(r = 45))+
  coord_cartesian(clip = 'off')+
  panel_border(color = 'black',size = 0.4)


```

##Fig. 1c
Copy number heatmap of chromosome 12 showing frequent 12p amplification and isochromosome 12p [i(12p)] in TGCTs.

```{r}
source('../functions/ggdendro.R')
load('../data/cnvdata.RData')
load('../data/isochromosome_12p.RData')
sample_order <- isochromosome_12p

reducedseg_chr12 <- reducedseg[reducedseg$chrom=='12',]
msegdata <- as.matrix(reducedseg_chr12[,-(1:3)])
msegdata <- apply(msegdata, 2, as.numeric)

hc <- hclust(dist(t(msegdata),method = 'euclidean'),method = 'ward.D')
hcdata <- dendro_data_k(hc, 3)
cnvclust <- as_tibble(hcdata$labels) %>% select(Tumor_Barcode=label,CNV_Clust=clust) %>% mutate(CNV_Clust=paste0('C',CNV_Clust))
chrlevels <- c(seq(1:22),"X")
p1 <- cnvdata %>% 
  mutate(chr=factor(chr,levels = chrlevels)) %>%
  mutate(startpos=as.integer(startpos),endpos=as.integer(endpos)) %>% 
  left_join(sample_order) %>% 
  arrange(Seq,chr) %>% 
  filter(chr %in% c(12)) %>% 
  ggplot(aes(fill=relative_copy))+
  #geom_segment(aes(x=startpos,xend=endpos,y=Tumor_Barcode,yend=Tumor_Barcode),size=3)+
  geom_rect(aes(xmin=startpos,xmax=endpos,ymin=Seq-0.5,ymax=Seq+0.5))+
  facet_grid(~chr,scales = 'free_x',space = 'free',switch="both")+
  scale_fill_gradient2(low = "blue",high = "red",mid = "white",midpoint = 0)+
  scale_x_continuous(breaks = pretty_breaks(),expand = c(0,0),sec.axis = dup_axis())+
  labs(fill="Relative Copy Number")+
  scale_y_continuous(breaks = sample_order$Seq,labels = sample_order$Tumor_Barcode,expand = expand_scale(add = c(0, 0)), sec.axis = dup_axis())+
  #facet_wrap(~chr,nrow = 1,scales = 'free_x',strip.position = "bottom")+
  labs(x="",y="")+
  #theme_ipsum_rc()+
  theme(text = element_text(family = "Roboto Condensed"),panel.spacing = unit(0, "lines"),axis.text.x = element_blank(),axis.ticks = element_blank(),axis.text.y = element_blank(),strip.placement = "outside",strip.switch.pad.wrap = unit(0, "cm"),strip.text = element_text(face = "bold",size=10),strip.background = element_rect(color = "gray50",fill="white"),panel.background = element_rect(fill = 'white'),axis.line= element_line(colour = 'black'),axis.title.y.right = element_blank(),axis.text.y.right = element_blank(),axis.ticks.y.right = element_blank(),axis.title.x.top = element_blank(),axis.text.x.top = element_blank(),axis.ticks.x.top = element_blank() )+
  coord_cartesian(clip="off")+
  geom_hline(yintercept = 30.5,linetype = 1)

## add subtype

p2 <- sample_order %>%
  left_join(wgs_data) %>% 
  mutate(chr=12) %>% 
  ggplot(aes(x=1,y=Seq,fill=Subtype))+
  geom_tile()+
  facet_grid(~chr,scales = 'free_x',space = 'free',switch="both")+
  scale_fill_manual(values = subtypecol)+
  scale_x_continuous(breaks = pretty_breaks(),expand = c(0,0),sec.axis = dup_axis())+
  labs(fill="Subtype")+
  scale_y_continuous(breaks = sample_order$Seq,labels = sample_order$Tumor_Barcode,expand = expand_scale(add = c(0, 0)), sec.axis = dup_axis())+
  #facet_wrap(~chr,nrow = 1,scales = 'free_x',strip.position = "bottom")+
  labs(x="",y="")+
  #theme_ipsum_rc()+
  theme_void()+
  theme(text = element_text(family = "Roboto Condensed"),panel.spacing = unit(0, "lines"),axis.text.x = element_blank(),axis.ticks = element_blank(),axis.text.y = element_blank(),strip.placement = "outside",strip.switch.pad.wrap = unit(0, "cm"),strip.text = element_text(face = "bold",size=10),strip.background = element_rect(color = "gray50",fill=NULL,linewidth = 0),panel.background = element_rect(fill = NULL,linewidth=0),axis.line= element_blank(),axis.title.y.right = element_blank(),axis.text.y.right = element_blank(),axis.ticks.y.right = element_blank(),axis.title.x.top = element_blank(),axis.text.x.top = element_blank(),axis.ticks.x.top = element_blank() )+
  coord_cartesian(clip="off")+
  geom_hline(yintercept = 30.5,linetype = 1)

p <- plot_grid(p1,p2,align = 'h',axis = 'tb',nrow = 1,rel_widths = c(9,2))

p

#ggsave(filename = 'heatmap_all_segements_isochromosome12p_raw.pdf',plot = p, width = 6.9,height = 10,device = cairo_pdf)

```

##Fig. 1d
Enrichment of i(12p) in seminomas with KIT mutations. P-value and odds ratio from two-sided Fisher’s exact test shown above the barplot.

```{r,fig.width=4.5}
# Association with KIT mutations ------------------------------------------
load('../data/tgct_maf.RData')

tdata <- wgs_data %>% 
  left_join(
    isochromosome_12p %>% select(Tumor_Barcode, isochromsome)  
  ) %>% 
  left_join(
    tgct_maf %>% filter(Hugo_Symbol == 'KIT') %>% select(Tumor_Barcode,aaChange) %>% mutate(KIT='Mutant')
  ) %>% 
  mutate(KIT = replace_na(KIT,'Wildtype'))


barplot_fisher(mdata0 = tdata %>% filter(Subtype=='Seminoma'),var1name = "isochromsome",var1lab = 'Isochromosome 12p', var2name = 'KIT',var2lab = 'KIT',Pcol = ncicolpal[1])

```

##Fig. 1e
GISTIC2.0 analysis of recurrent focal copy number alterations. Significant amplifications (red) and deletions (blue) are annotated with potential target genes. Y-axis represents the significance by q value (left) and normalized amplification signals (G-Score; right). The green line represents the significance cutoff at Q value=0.25.

```{r,fig.width=12}

load('../data/Gistic_gcp.RData')
# Gistic plot -------------------------------------------------------------

cyto <- read_delim('../data/cytoBand.txt.gz',delim = '\t',col_names = F)
colnames(cyto) <- c('Chromosome', 'Start' , 'End', 'Band', 'gieStain')
cyto <- cyto %>% mutate(Chromosome = str_remove(Chromosome,"chr"),Centro=if_else(gieStain=="acen",1,0), Arm=paste0(Chromosome,str_sub(Band,1,1))) %>% filter(Chromosome %in% c(1:22,"X","Y"))
#hg38centro <- cyto %>% filter(Centro!=0,Chromosome %in% c(1:22)) %>% select(chrom=Chromosome,start=Start,end=End,Centro) 

hg38centro <- cyto %>% 
  filter(Centro!=0,Chromosome %in% c(1:22)) %>% 
  select(Chromosome,Start_Position=Start,End_Position=End) %>% 
  group_by(Chromosome) %>% 
  arrange(End_Position) %>% 
  slice(1) %>% 
  select(Chromosome,Centro=End_Position) %>% 
  ungroup()



chrlevels <- as.character(c(1:23,"X"))

hg38backgroup <- 
  cyto %>% group_by(Chromosome) %>% summarise(Start=min(Start)+1,End=max(End)) %>% 
  filter(Chromosome %in% c(1:23,"X")) %>% mutate(Chromosome=factor(Chromosome,levels = chrlevels))



# Amplificaiton -----------------------------------------------------------

ampdata <- bind_rows(
  gcp$gis.scores %>% filter(Variant_Classification=="Amp") %>% select(Chromosome,pos=Start_Position,G_Score,fdr) %>% mutate(Chromosome = if_else(Chromosome == 23,'X',Chromosome)),
  gcp$gis.scores %>% filter(Variant_Classification=="Amp") %>% select(Chromosome,pos=End_Position ,G_Score,fdr) %>% mutate(Chromosome = if_else(Chromosome == 23,'X',Chromosome)),
  hg38centro %>% mutate(fdr=0,G_Score=0) %>% select(Chromosome,pos=Centro,fdr,G_Score),
  hg38backgroup %>% mutate(fdr=0,G_Score=0) %>%  select(Chromosome,pos=Start,fdr,G_Score),
  hg38backgroup %>% mutate(fdr=0,G_Score=0) %>%  select(Chromosome,pos=End,fdr,G_Score)
) %>% arrange(pos) %>% 
  filter(Chromosome %in% c(1:22,"X")) 

fdrmax <- max(ampdata$fdr)
ampdata_limit <- ampdata %>% group_by(Chromosome) %>% summarise(min=min(pos),max=max(pos)) %>% mutate(fdr_max=150,pos=NA_integer_) %>% mutate(Chromosome=factor(Chromosome,levels = chrlevels)) %>% arrange(Chromosome)

ampdata_limit$col <- rep(c('white','gray90'),12)[1:23]

secbreacks <- NULL
secbreacks[1] <- gcp$gis.scores %>% filter(Variant_Classification=="Amp") %>% arrange(G_Score) %>% filter(G_Score>=0.9) %>% slice(1) %>% pull(fdr)
secbreacks[2] <- gcp$gis.scores %>% filter(Variant_Classification=="Amp") %>% arrange(G_Score) %>% filter(G_Score>=0.8) %>% slice(1) %>% pull(fdr)
secbreacks[3] <-gcp$gis.scores %>% filter(Variant_Classification=="Amp") %>% arrange(G_Score) %>% filter(G_Score>=0.6) %>% slice(1) %>% pull(fdr)
secbreacks[4] <-gcp$gis.scores %>% filter(Variant_Classification=="Amp") %>% arrange(G_Score) %>% filter(G_Score>=0.4) %>% slice(1) %>% pull(fdr)
secbreacks[5] <-gcp$gis.scores %>% filter(Variant_Classification=="Amp") %>% arrange(G_Score) %>% filter(G_Score>=0.2) %>% slice(1) %>% pull(fdr)
secbreacks[6] <- -log10(0.25)
names(secbreacks) <- c(0.9,0.8,0.6,0.4,0.2,0)
names(secbreacks)[6] <- gcp$gis.scores %>% filter(Variant_Classification=="Amp") %>% arrange(fdr) %>% filter(fdr>= -log10(0.25)) %>% slice(1) %>% pull(G_Score) %>% round(digits = 2)

p_amp <- ampdata %>% 
  ggplot()+
  geom_rect(data=ampdata_limit,aes(xmin=min,xmax=max,ymin=0,ymax=320),fill=ampdata_limit$col)+
  geom_step(aes(pos,fdr),size=0.5,colour="red",na.rm = TRUE,direction = 'vh')+
  facet_grid(~factor(Chromosome,levels = chrlevels),scales = "free_x",space = 'free')+
  geom_vline(data=hg38centro,aes(xintercept=Centro),size=0.2,colour="gray20",linetype=4)+
  geom_hline(data=hg38centro %>% mutate(fdrcutoff=2),aes(yintercept=fdrcutoff),size=0.2,colour="green4",linetype=1)+
  scale_x_continuous(breaks = pretty_breaks(),expand = c(0,0),sec.axis = dup_axis())+
  scale_y_continuous(breaks = c(1,5*c(1,2,4,8,16,32,64)),expand = expand_scale(add = c(0, 0)), sec.axis = dup_axis(breaks = secbreacks,labels = names(secbreacks),name = 'G-Score'),labels = math_format(10^-.x),trans='log2')+
  labs(x="",y="q-value")+
  theme(text = element_text(family = "Roboto Condensed"),panel.spacing = unit(0, "lines"),panel.grid.major.y = element_line(colour = 'gray50',linetype = 5),axis.text.x = element_blank(),axis.ticks.x = element_blank(),axis.text.y = element_text(size = 12),axis.title.y = element_text(size = 14),strip.placement = "inside",strip.switch.pad.wrap = unit(0, "cm"),strip.text = element_blank(),strip.background = element_blank(),axis.line= element_line(colour = 'black'),axis.title.y.right = element_text(size=14),axis.text.y.right = element_text(size = 12),axis.title.x.top = element_blank(),axis.text.x.top = element_blank(),axis.ticks.x.top = element_blank() ) +
  panel_border(colour = 'gray80',size = 0.2)+coord_cartesian(ylim = c(0.5,320))


#focal_cnv_amp <- focal_cnv %>% filter(Variant_Classification=="Amp") %>% mutate(pos=(Start_Position+End_Position)/2,fdr=-log10(qvalues))%>% select(Chromosome,pos,fdr,Cytoband,Drive_Gene,TotalGene) %>% mutate(gene=Cytoband) %>% mutate(Chromosome=factor(Chromosome,levels = chrlevels))

# focal_cnv_amp$gene[focal_cnv_amp$Cytoband=="12q15"]="\n12q15/MDM2"
# focal_cnv_amp$gene[focal_cnv_amp$Cytoband=="5p15.33"]="5p15.33/TERT"
# focal_cnv_amp$gene[focal_cnv_amp$Cytoband=="8q24.21"]="8q24.21/MYC\n"
# focal_cnv_amp$gene[focal_cnv_amp$Cytoband=="14q13.3"]="14q13.3/NKX2-1"
# focal_cnv_amp$gene[focal_cnv_amp$Cytoband=="16p13.2"]="16p13.2/GRIN2A"
# focal_cnv_amp$gene[focal_cnv_amp$Cytoband=="1q21.3"]="1q21.3/MCL1"
# focal_cnv_amp$gene[focal_cnv_amp$Cytoband=="7q36.1"]="7q36.1/BRAF"
# #focal_cnv_amp$gene[focal_cnv_amp$Cytoband=="16q12.1"]="16q12.1/N4BP1"
# 
# focal_cnv_amp$gene[focal_cnv_amp$Cytoband=="20q11.23"]="20q11.23"
# focal_cnv_amp$gene[focal_cnv_amp$Cytoband=="20q13.33"]="20q13.33  " 
# 
# focal_cnv_amp$gene[focal_cnv_amp$Cytoband=="1p34.2"]=""
# focal_cnv_amp$gene[focal_cnv_amp$Cytoband=="3q26.2"]=""
# focal_cnv_amp$gene[focal_cnv_amp$Cytoband=="13q33.3"]=""
# focal_cnv_amp$gene[focal_cnv_amp$Cytoband=="16q12.1"]=""
# focal_cnv_amp$gene[focal_cnv_amp$Cytoband=="17q24.3"]=""
# focal_cnv_amp <- bind_rows(
#   focal_cnv_amp,
#   tibble(Chromosome=factor(7,levels = chrlevels),pos=55086725,fdr=2,gene="7p11.2/EGFR")
# )






# p_amp <- p_amp +geom_text_repel(
#   data = focal_cnv_amp,
#   aes(x=pos,y=fdr,label=gene),
#   nudge_y      = 2,
#   direction    = "x",
#   angle        = 90,
#   vjust        = 0.5,
#   segment.size = 0,
#   segment.color = 'black',
#   force = 10,
#   size=3,
#   fontface="bold",
#   family="Roboto Condensed"
# )

#p_amp


# Deletion ----------------------------------------------------------------

reverselog_trans <- function(base = exp(1)) {
  trans <- function(x) -log(x, base)
  inv <- function(x) base^(-x)
  trans_new(paste0("reverselog-", format(base)), trans, inv, 
            log_breaks(base = base), 
            domain = c(1e-100, Inf))
}


deldata <- bind_rows(
  gcp$gis.scores %>% filter(Variant_Classification=="Del") %>% select(Chromosome,pos=Start_Position,G_Score,fdr) %>% mutate(Chromosome = if_else(Chromosome == 23,'X',Chromosome)),
  gcp$gis.scores %>% filter(Variant_Classification=="Del") %>% select(Chromosome,pos=End_Position ,G_Score,fdr) %>% mutate(Chromosome = if_else(Chromosome == 23,'X',Chromosome)),
  hg38centro %>% mutate(fdr=0,G_Score=0) %>% select(Chromosome,pos=Centro,fdr,G_Score),
  hg38backgroup %>% mutate(fdr=0,G_Score=0) %>%  select(Chromosome,pos=Start,fdr,G_Score),
  hg38backgroup %>% mutate(fdr=0,G_Score=0) %>%  select(Chromosome,pos=End,fdr,G_Score)
) %>% arrange(pos) %>% 
  filter(Chromosome %in% c(1:22,"X")) 

fdrmax <- max(deldata$fdr)
deldata_limit <- deldata %>% group_by(Chromosome) %>% summarise(min=min(pos),max=max(pos)) %>% mutate(fdr_max=35,pos=NA_integer_) %>% mutate(Chromosome=factor(Chromosome,levels = chrlevels)) %>% arrange(Chromosome)

ampdata_limit$col <- rep(c('white','gray90'),12)[1:23]

secbreacks <- NULL
#secbreacks[1] <- gcp$gis.scores %>% filter(Variant_Classification=="Del") %>% arrange(G_Score) %>% filter(G_Score>=2.0) %>% slice(1) %>% pull(fdr)
secbreacks[1] <- 300
#secbreacks[2] <- gcp$gis.scores %>% filter(Variant_Classification=="Del") %>% arrange(G_Score) %>% filter(G_Score>=1.6) %>% slice(1) %>% pull(fdr)
secbreacks[2] <- 210
secbreacks[3] <-gcp$gis.scores %>% filter(Variant_Classification=="Del") %>% arrange(G_Score) %>% filter(G_Score>=0.8) %>% slice(1) %>% pull(fdr)
secbreacks[4] <-gcp$gis.scores %>% filter(Variant_Classification=="Del") %>% arrange(G_Score) %>% filter(G_Score>=0.4) %>% slice(1) %>% pull(fdr)
secbreacks[5] <- -log10(0.25)
names(secbreacks) <- c(2.0,1.6,0.8,0.4,0)
names(secbreacks)[5] <- gcp$gis.scores %>% filter(Variant_Classification=="Del") %>% arrange(fdr) %>% filter(fdr>=-log10(0.25)) %>% slice(1) %>% pull(G_Score) %>% round(digits = 2)



p_del <- deldata %>% 
  ggplot()+
  geom_rect(data=deldata_limit,aes(xmin=min,xmax=max,ymin=0,ymax=320),fill=ampdata_limit$col)+
  geom_step(aes(pos,fdr),size=0.5,colour="blue",na.rm = TRUE,direction = 'vh')+
  facet_grid(~factor(Chromosome,levels = chrlevels),scales = "free_x",space = 'free')+
  geom_vline(data=hg38centro,aes(xintercept=Centro),size=0.2,colour="gray20",linetype=4)+
  geom_hline(data=hg38centro %>% mutate(fdrcutoff=2),aes(yintercept=fdrcutoff),size=0.2,colour="green4",linetype=1)+
  scale_x_continuous(breaks = pretty_breaks(),expand = expansion(mult = c(0,0.02)),sec.axis = dup_axis())+
  scale_y_continuous(breaks = 5*c(1,2,4,8,16,32,64),expand = expand_scale(add = c(0, 0)), sec.axis = dup_axis(breaks = secbreacks,labels = names(secbreacks),name = 'G-Score'),labels = math_format(10^-.x),trans=reverselog_trans(2))+
  labs(x="",y="q-value")+
  theme(text = element_text(family = "Roboto Condensed"),panel.spacing = unit(0, "lines"),panel.grid.major.y = element_blank(),panel.background = element_blank(),axis.text.x = element_blank(),axis.ticks.x = element_blank(),axis.text.y = element_text(size = 12),axis.title.y = element_text(size = 14),strip.placement = "inside",strip.switch.pad.wrap = unit(0, "cm"),strip.text = element_blank(),strip.background = element_blank(),axis.line= element_line(colour = 'black'),axis.title.y.right = element_text(size=14),axis.text.y.right = element_text(size = 12),axis.title.x.top = element_blank(),axis.text.x.top = element_blank(),axis.ticks.x.top = element_blank() )+
  panel_border(colour = 'gray80',size = 0.2)
#coord_cartesian(ylim = c(0.2,320))


#focal_cnv_del <- focal_cnv %>% filter(Variant_Classification=="Del") %>% mutate(pos=(Start_Position+End_Position)/2,fdr=-log10(qvalues))%>% select(Chromosome,pos,fdr,Cytoband,Drive_Gene,TotalGene) %>% mutate(gene=Cytoband) %>% mutate(Chromosome=factor(Chromosome,levels = chrlevels))
# #focal_cnv_del %>% View()
# 
# focal_cnv_del$gene[focal_cnv_del$Cytoband=="9p21.3"]="\n9p21.3/CDKN2A"
# focal_cnv_del$gene[focal_cnv_del$Cytoband=="8p23.3"]="8p23.3/CSMD1"
# focal_cnv_del$gene[focal_cnv_del$Cytoband=="19p13.3"]="19p13.3/STK11\n"
# 
# focal_cnv_del$gene[focal_cnv_del$Cytoband=="3q27.3"]=""
# 
# focal_cnv_del <- bind_rows(
#   focal_cnv_del,
#   tibble(Chromosome=factor(12,levels = chrlevels),pos=6055366,fdr=3,gene="12p13.1")
# )

# p_del <- p_del +geom_text_repel(
#   data = focal_cnv_del,
#   aes(x=pos,y=fdr,label=gene),
#   nudge_y      = 2,
#   direction    = "x",
#   angle        = 90,
#   vjust        = 3,
#   segment.size = 0,
#   segment.color = 'black',
#   force = 10,
#   size=3,
#   fontface="bold",
#   family="Roboto Condensed"
# )

#p_del

# Cytoband plot -----------------------------------------------------------

p_cytoband <- gcp$chr.tbl %>%
  mutate(pos=(start+end)/2) %>% 
  filter(chr %in% c(1:23)) %>% 
  ggplot()+
  geom_rect(aes(xmin=start,xmax=end,ymin=0.2,ymax=0.8,fill=color),color="black")+
  geom_text(aes(x=pos,y = 0.5,label=chr),size=3,fontface = "bold",family="Roboto Condensed")+
  theme_void()+
  scale_fill_manual(values = c('white','gray50'))+
  theme(legend.position = 'none',text = element_text(family = "Roboto Condensed"))+
  scale_x_continuous(breaks = pretty_breaks(),expand = c(0,0))

# Combined Plots ----------------------------------------------------------
pdfhr()
pdfhr2()
#extrafont::font_import()
p_combined_plot <- cowplot::plot_grid(
  p_amp+theme(plot.margin=margin(b=-0.45,unit="cm")),
  p_cytoband+theme(plot.margin=margin(b=0,unit="cm")),
  p_del+theme(plot.margin=margin(t=0,unit="cm")),
  align = 'v',axis = "lr",rel_heights = c(2,0.1,2),ncol = 1)

p_combined_plot

#ggsave(file="Gistic2_ztw.pdf",plot = p_combined_plot,width = 12,height = 6,device = cairo_pdf)

```

##Fig. 1f
Oncoprints depicting recurrent driver mutations, focal amplifications/deletions, and genes disrupted by SVs in seminomas and non-seminomas. Mutation frequencies for each gene are shown on the right. For mutations, only recurrent nonsynonymous mutations in known pan-cancer driver genes were included. For focal somatic copy number alterations (SCNAs), only high-level amplifications (copy number >= 2) and deep deletions (copy number = –2) were considered. For genes disrupted by SVs, we included only recurrent events involving known pan-cancer driver genes.
All boxplots display median, interquartile range (IQR), and whiskers extending to 1.5× IQR.

```{r,fig.width=12}
conflicts_prefer(tidylog::mutate)
conflicts_prefer(dplyr::left_join)
conflicts_prefer(dplyr::count)
conflicts_prefer(dplyr::ungroup)
conflicts_prefer(dplyr::summarise)
conflicts_prefer(dplyr::group_by)
conflicts_prefer(dplyr::rename)
conflicts_prefer(tidyr::pivot_wider)
conflicts_prefer(dplyr::slice)

# presetting
library(data.table)
conflicts_prefer(cowplot::get_legend)
source('../functions/Oncoplots_functions.R')
tmp <- read_csv('../functions/oncoplot_colors.csv')
landscape_colors <- tmp$Color
names(landscape_colors) <- tmp$Name
#landscape_colors <- c(landscape_colors,sp_group_color_new)

landscape_colors <- c(
  landscape_colors,
  c('Multiple WGD' = "#14315C", 'EUR_Pop' = "#fc8d62", 'Mix_Pop' = "#542788")
)


# load Genomic alterations data -------------------------------------------------
load('../data/wgs_data.RData', verbose = T)
# driver genes
load('../data/tgct_maf.RData', verbose = T)
load('../data/intogene_drivers.RData', verbose = T)
load('../data/DriveGene.RData')


genelist <- tgct_maf_freq %>%
  filter(
    Hugo_Symbol %in% c(drivegene$Gene, intogene$SYMBOL, 'SRCAP'),
    nMutSample > 2
  ) %>%
  pull(Hugo_Symbol)


data_mutation <- tgct_maf %>%
  filter(Hugo_Symbol %in% genelist) %>%
  left_join(wgs_data %>% select(Subject, Tumor_Barcode)) %>%
  mutate(Alteration = Variant_Classification) %>%
  #mutate(Alteration=paste0(Variant_Classification,":",tx,":",aaChange)) %>%
  select(Subject, Tumor_Barcode, Gene = Hugo_Symbol, Alteration) %>%
  mutate(Type = 'Mutation_Driver') %>%
  unique()


# focal amplification
load('../data/tgct_gistic_gene.RData')
## need to define the gene set ##
genelist2 <- c(
  genelist,
  c(
    'GRID2',
    'CHEK1',
    'FANCA',
    'SPRY3',
    'KIT',
    'HIP1',
    'CDKN1B',
    'MDM2',
    'MAPK1'
  )
) %>%
  unique()

data_gistic <- tgct_gistic_gene %>%
  mutate(GeneSymbol = str_remove(GeneSymbol, "\\|.*")) %>%
  filter(GeneSymbol %in% genelist2)

data_gistic <- data_gistic %>%
  filter(CNV %in% c(-2, 2)) %>%
  mutate(CNV = if_else(CNV == 2, "Amp", "Del")) %>%
  left_join(wgs_data %>% select(Subject, Tumor_Barcode)) %>%
  mutate(Gene = GeneSymbol, Alteration = CNV, Type = "SCNA_Focal_Gene") %>%
  select(Subject, Tumor_Barcode, Gene, Alteration, Type)

# Fusion
load('../data/SV_Fusion.RData', verbose = T)
genelist3 <- c(genelist, c('JARID2', 'PARK2', 'GRID2'))
data_fusion <- data_fusion %>% filter(Gene %in% genelist3)

## save R object
data_top <-
  bind_rows(
    data_mutation,
    data_gistic,
    data_fusion
  )

#save(data_top, file = 'data_top.RData')

## combined all together
data_top <-
  bind_rows(
    data_mutation %>% mutate(Type = 'Alterations'),
    data_gistic %>% mutate(Type = 'Alterations'),
    data_fusion %>% mutate(Type = 'Alterations'),
  )


# Add data feature
load('../data/MCN_data.RData', verbose = T)
load('../data/VerifyBamID_idata.RData')
load('../data/isochromosome_12p.RData')

data_feature1 <- idata %>%
  select(Subject, WGS_Inferred_Ancestry = POP) %>%
  left_join(wgs_data) %>%
  mutate(Gene = 'Ancestry', Type = 'Feature') %>%
  select(Subject, Tumor_Barcode, Gene, Alteration = WGS_Inferred_Ancestry, Type)

data_feature2 <- MCN_data %>%
  mutate(
    Alteration = if_else(WGD2 == 'WGD2', 'Multiple WGD', WGD),
    Gene = 'WGD',
    Type = 'Feature'
  ) %>%
  select(Subject, Tumor_Barcode, Gene, Alteration, Type)

data_feature3 <- isochromosome_12p %>%
  select(Tumor_Barcode, isochromsome) %>%
  left_join(wgs_data) %>%
  mutate(Gene = 'i(12p)', Type = 'Feature') %>%
  select(Subject, Tumor_Barcode, Gene, Alteration = isochromsome, Type)


data_feature <- data_feature1 %>%
  bind_rows(data_feature2) %>%
  bind_rows(data_feature3)

#
# data_feature <- sp_group_data2 %>%
#   filter(Tumor_Barcode %in% kras_samples) %>%
#   mutate(Type='Genomic_Feature',Gene='Group') %>%
#   left_join(wgs_groups_info) %>%
#   select(Subject,Tumor_Barcode,Gene,Alteration = SP_Group_New,Type)

## oncoplot

for(subtype0 in c('Seminoma','Non-Seminoma')){
  
  sample_level0 <- wgs_data %>%
    filter(Subtype == subtype0) %>%
    pull(Tumor_Barcode)
  
  
  # select only recurrent events --------------------------------------------
  tmp <- data_mutation %>%
    filter(Tumor_Barcode %in% sample_level0) %>%
    select(Subject, Gene) %>%
    unique() %>%
    count(Gene, sort = T) %>%
    filter(n > 1) %>%
    pull(Gene)
  
  data_mutation <- data_mutation %>% filter(Gene %in% tmp)
  
  tmp <- data_gistic %>%
    filter(Tumor_Barcode %in% sample_level0) %>%
    select(Subject, Gene) %>%
    unique() %>%
    count(Gene, sort = T) %>%
    filter(n > 1) %>%
    pull(Gene)
  
  data_gistic <- data_gistic %>% filter(Gene %in% tmp)
  
  tmp <- data_fusion %>%
    filter(Tumor_Barcode %in% sample_level0) %>%
    select(Subject, Gene) %>%
    unique() %>%
    count(Gene, sort = T) %>%
    filter(n > 1) %>%
    pull(Gene)
  
  data_fusion <- data_fusion %>% filter(Gene %in% tmp)
  
  
  result_mutation <- oncoplot(
    data = data_mutation,
    landscape_colors = landscape_colors,
    sample_level0 = sample_level0,
    GeneSortOnly = TRUE
  )
  
  
  result_gistic <- oncoplot(
    data = data_gistic,
    landscape_colors = landscape_colors,
    sample_level0 = sample_level0,
    GeneSortOnly = TRUE
  )
  
  #
  # result_fusion <- oncoplot(
  #   data = data_fusion,
  #   landscape_colors = landscape_colors,
  #   sample_level0 = sample_level0,
  #   GeneSortOnly = TRUE
  # )
  
  result_feature <- oncoplot(
    data = data_feature,
    landscape_colors = landscape_colors,
    sample_level0 = sample_level0,
    GeneSortOnly = TRUE
  )
  
  sample_new_level <- result_mutation$sample_level %>%
    left_join(
      result_gistic$sample_level
    ) %>%
    left_join(
      result_feature$sample_level
    ) %>%
    arrange(Mutation_Driver, SCNA_Focal_Gene, Feature) %>%
    pull(Tumor_Barcode)
  
  
  result_mutation <- oncoplot(
    data = data_mutation,
    landscape_colors = landscape_colors,
    sample_level0 = sample_level0,
    sample_level = sample_new_level,
    tmar = -0.2,
    bmar = -0.05
  )
  result_gistic <- oncoplot(
    data = data_gistic,
    landscape_colors = landscape_colors,
    sample_level0 = sample_level0,
    sample_level = sample_new_level,
    tmar = -0.2,
    bmar = -0.05
  )
  # result_fusion <- oncoplot(
  #   data = data_fusion,
  #   landscape_colors = landscape_colors,
  #   sample_level0 = sample_level0,
  #   sample_level = sample_new_level,
  #   GeneSortOnly = TRUE,
  #   bmar = 0,
  #   p2_axis_hidden = TRUE,
  #   p2_hidden = TRUE,
  #   removeAlterbyColor = TRUE
  # )
  
  result_feature <- oncoplot(
    data = data_feature,
    landscape_colors = landscape_colors,
    sample_level0 = sample_level0,
    sample_level = sample_new_level,
    GeneSortOnly = TRUE,
    bmar = 0,
    p2_axis_hidden = TRUE,
    p2_hidden = TRUE,
    removeAlterbyColor = FALSE
  )
  
  
  oncoplot_final <- oncoplot_combined(
    result_mutation,
    result_gistic,
    result_feature
  )
  
  print(paste0(subtype0,'\n\n'))
  print(oncoplot_final)
}


```



## R session information
```{r}
sessionInfo()
```
