---
title: "A Thiopurine-like Mutagenic Process Defines TGCT Subtypes"
author: "Tongwu Zhang"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

## Overview
This R Markdown document reproduces the main analyses and figures for TCGA-TGCT WGS dataset

Manuscript title: A Thiopurine-like Mutagenic Process Defines TGCT Subtype

Fig. 2: Widespread Chromosome X Amplification and X Chromosome Inactivation (XCI) in TGCTs.
---

## Load Required Libraries and Set Plotting Styles 
(You may need to install some packages if missing)

```{r}

source('../functions/Sherlock_functions.R')

set_wd()
libztw()
pdfhr()
pdfhr2()

# Load R object WGS analysis results --------------------------------------

load('../data/wgs_data.RData')
load('../data/wgs_clinical.RData')
load('../data/ngspurity_data.RData')


```
##Fig. 2a
Tumor purity-adjusted sequencing depth ratio of chromosome X versus autosomes in matched normal blood, seminomas, and non-seminomas. P-values from two-sided Wilcoxon rank-sum tests are shown above boxplots.


```{r,fig.width=3}

load('../data/Gender.RData', verbose = T)

my_comparisons = list(
  c('Seminoma', 'Normal'),
  c('Seminoma', 'Non-Seminoma'),
  c('Non-Seminoma', 'Normal')
)

pdata <- tdata %>%
  mutate(Type = if_else(Type == 'Blood Derived Normal', 'Normal', 'Tumor')) %>%
  filter(Barcode %in% c(wgs_data$Tumor_Barcode, wgs_data$Normal_Barcode)) %>%
  left_join(wgs_data %>% select(Barcode = Tumor_Barcode, Subtype)) %>%
  mutate(Type = if_else(Type == 'Tumor', Subtype, Type)) %>%
  mutate(Type = factor(Type, levels = c('Normal', 'Seminoma', 'Non-Seminoma')))

stat.test <- pdata %>%
  rstatix::wilcox_test(RatioX ~ Type, comparisons = my_comparisons) %>%
  rstatix::add_xy_position(x = "Type") %>%
  mutate(myformatted.p = paste0('P = ', scientific_format(digits = 3)(p))) %>%
  mutate(y.position = y.position - 0.1) %>%
  mutate(Type = NA)

pdata %>%
  ggplot(aes(Type, RatioX)) +
  geom_hline(yintercept = 0.5, col = '#cccccc', linetype = 2, linewidth = 0.5) +
  geom_point(
    aes(fill = Type),
    pch = 21,
    size = 2,
    position = position_jitter(width = 0.2, height = 0),
    stroke = 0.2,
    col = 'white'
  ) +
  geom_boxplot(
    width = 0.6,
    outlier.shape = NA,
    fill = NA,
    col = ncicolpal[1],
    size = 0.6
  ) +
  theme_ipsum_rc(axis_title_just = 'm', axis_title_size = 14, ticks = T) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    panel.spacing = unit(0.2, "lines"),
    strip.background = element_rect(fill = 'gray95'),
    strip.text = element_text(hjust = 0.5)
  ) +
  theme(panel.spacing = unit(0.3, 'cm'), plot.margin = margin(4, 4, 4, 4)) +
  labs(x = NULL, y = 'ChrX depth / Autosome depth') +
  scale_y_continuous(breaks = pretty_breaks(n = 6), limits = c(NA, 1.25)) +
  scale_fill_manual(values = c('Normal' = "#3D4551", subtypecol)) +
  guides(fill = "none") +
  panel_border(color = 'black', linetype = 1) +
  stat_pvalue_manual(stat.test, label = "myformatted.p", color = ncicolpal[1])

#ggsave(file = 'Sex_RatioX.pdf', width = 3, height = 5.5, device = cairo_pdf)

load('../data/ngspurity_data.RData')
load('../data/wgs_metrics.RData')

pdata %>%
  filter(Type != 'Normal') %>%
  left_join(ngspurity %>% select(Barcode = Tumor_Barcode, BB_Purity)) %>%
  do(tidy(lm(RatioX ~ autosome_depth + BB_Purity + Type, data = .)))

# adjust by the tumor purity

pdata2 <- pdata %>%
  left_join(wgs_metrics %>% select(Barcode = Sample_Barcode, MEDIAN_COVERAGE))

pdata2 <- left_join(
  pdata2 %>%
    filter(Type != 'Normal') %>%
    select(
      Subject,
      Subtype,
      Tumor_Barcode = Barcode,
      Tumor_chrX = chrX,
      Tumor_chrY = chrY,
      Tumor_autosome = autosome_depth,
      Tumor_cov = MEDIAN_COVERAGE
    ),
  pdata2 %>%
    filter(Type == 'Normal') %>%
    select(
      Subject,
      Normal_Barcode = Barcode,
      Normal_chrX = chrX,
      Normal_chrY = chrY,
      Normal_autosome = autosome_depth,
      Normal_cov = MEDIAN_COVERAGE
    )
) %>%
  left_join(
    ngspurity %>% select(Tumor_Barcode, BB_Purity)
  ) %>%
  mutate(
    Tumor_chrX_adjust = Tumor_chrX -
      (1 - BB_Purity) * (Normal_chrX * Tumor_cov / Normal_cov)
  ) %>%
  mutate(
    Tumor_chrY_adjust = Tumor_chrY -
      (1 - BB_Purity) * (Normal_chrY * Tumor_cov / Normal_cov)
  ) %>%
  mutate(
    Tumor_autosome_adjust = Tumor_autosome -
      (1 - BB_Purity) * (Normal_autosome * Tumor_cov / Normal_cov)
  ) %>%
  mutate(RatioX = Tumor_chrX_adjust / Tumor_autosome_adjust) %>%
  mutate(RatioY = Tumor_chrY_adjust / Tumor_autosome_adjust)

pdata3 <- bind_rows(
  pdata2 %>%
    select(Subject, Type = Subtype, Barcode = Tumor_Barcode, RatioX, RatioY),
  pdata2 %>%
    mutate(
      Subtype = 'Normal',
      RatioX = Normal_chrX / Normal_autosome,
      RatioY = Normal_chrY / Normal_autosome
    ) %>%
    select(Subject, Type = Subtype, Barcode = Tumor_Barcode, RatioX, RatioY)
) %>%
  mutate(Type = factor(Type, levels = c('Normal', 'Seminoma', 'Non-Seminoma')))


stat.test <- pdata3 %>%
  rstatix::wilcox_test(RatioX ~ Type, comparisons = my_comparisons) %>%
  rstatix::add_xy_position(x = "Type") %>%
  mutate(myformatted.p = paste0('P = ', scientific_format(digits = 3)(p))) %>%
  mutate(y.position = y.position - 0.6) %>%
  mutate(Type = NA)

pdata3 %>%
  ggplot(aes(Type, RatioX)) +
  geom_hline(yintercept = 0.5, col = '#cccccc', linetype = 2, linewidth = 0.5) +
  geom_point(
    aes(fill = Type),
    pch = 21,
    size = 2,
    position = position_jitter(width = 0.2, height = 0),
    stroke = 0.2,
    col = 'white'
  ) +
  geom_boxplot(
    width = 0.6,
    outlier.shape = NA,
    fill = NA,
    col = ncicolpal[1],
    size = 0.6
  ) +
  theme_ipsum_rc(axis_title_just = 'm', axis_title_size = 14, ticks = T) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    panel.spacing = unit(0.2, "lines"),
    strip.background = element_rect(fill = 'gray95'),
    strip.text = element_text(hjust = 0.5)
  ) +
  theme(panel.spacing = unit(0.3, 'cm'), plot.margin = margin(4, 4, 4, 4)) +
  labs(x = NULL, y = 'Purity adjusted sequencing ratio (chrX/Autosome)') +
  scale_y_continuous(breaks = pretty_breaks(n = 6), limits = c(NA, 1.8)) +
  scale_fill_manual(values = c('Normal' = "#3D4551", subtypecol)) +
  guides(fill = "none") +
  panel_border(color = 'black', linetype = 1) +
  stat_pvalue_manual(stat.test, label = "myformatted.p", color = ncicolpal[1])

```

##Fig. 2b
Copy number profiles of chromosome X in TGCTs, separated by seminoma (left) and non-seminoma (right), ordered by total copy number. The y-axis indicates chromosome X positions. Top panel: clonal somatic copy number alterations (SCNAs); bottom panel: subclonal SCNAs. Subclonal fractions per sample are shown in the barplot below each panel.


```{r,fig.width=14,fig.height=12}

library(ggnewscale)

load('../data/wgs_data.RData', verbose = T)
load('../data/ngspurity_data.RData', verbose = T)

scna_chrx <- read_delim('../data/BBprofile_final_chrX.txt',delim = '\t',col_names = T)

# switch the nMaj and nMinor
tmp <- scna_chrx %>%
  mutate(size = abs(endpos - startpos)) %>%
  group_by(Tumor_Barcode) %>%
  arrange(desc(size)) %>%
  slice(1) %>%
  ungroup() %>%
  left_join(wgs_data) %>%
  arrange(Subtype, desc(nMaj1)) %>%
  select(Tumor_Barcode, Subtype) %>%
  mutate(seq = seq_along(Tumor_Barcode))

scna_chrx <- left_join(scna_chrx, tmp)

scna_color <- c(pal_gsea()(12)[c(2, 6:12)], 'darkred')
names(scna_color) <- as.character(c(0:8))

p1 <- scna_chrx %>%
  #filter(Tumor_Barcode == 'TCGA-2G-AAEQ-01A-11D-A734-36') %>%
  ggplot() +
  geom_rect(aes(
    xmin = seq - 0.5,
    xmax = seq + 0.5,
    ymin = startpos / 1e6,
    ymax = endpos / 1e6,
    fill = as.factor(nMaj1)
  )) +
  scale_fill_manual(breaks = 0:8, values = scna_color) +
  #geom_hline(yintercept = -1e4)+
  labs(x = NULL, y = 'Chromosome X (Mbp)', fill = 'SCNA') +
  guides(fill = guide_legend(nrow = 1)) +
  new_scale_fill() +
  geom_rect(aes(
    xmin = seq - 0.5,
    xmax = seq + 0.5,
    ymin = -5,
    ymax = 0,
    fill = frac1
  )) +
  scale_fill_viridis_c(direction = -1, limits = c(0, 1)) +
  scale_y_continuous(
    breaks = pretty_breaks(n = 7),
    limits = c(-5, 156040895 / 1e6),
    expand = c(0, 0)
  ) +
  scale_x_continuous(breaks = pretty_breaks(n = 7), expand = c(0, 0)) +
  theme_ipsum_rc(
    base_size = 13,
    axis_title_just = 'm',
    axis_title_size = 15,
    plot_margin = margin(5.5, 5.5, 5.5, 5.5),
    ticks = T,
    grid = F
  ) +
  theme(
    panel.spacing.x = unit(0.1, 'cm'),
    strip.text.x = element_text(hjust = 0.5, face = 'bold'),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    legend.position = "top",
    legend.direction = 'horizontal',
    legend.key.width = unit(2, 'cm')
  ) +
  panel_border(color = '#cccccc', size = 0.8) +
  facet_wrap(~Subtype, scales = 'free_x') +
  labs(fill = 'Clonal Fraction\n')

p2 <- scna_chrx %>%
  mutate(nMaj2 = if_else(frac2 == 0, NA_real_, nMaj2)) %>%
  #filter(Tumor_Barcode == 'TCGA-2G-AAEQ-01A-11D-A734-36') %>%
  ggplot() +
  geom_rect(aes(
    xmin = seq - 0.5,
    xmax = seq + 0.5,
    ymin = startpos / 1e6,
    ymax = endpos / 1e6,
    fill = as.factor(nMaj2)
  )) +
  scale_fill_manual(breaks = c(0:8), values = scna_color, na.value = 'white') +
  #geom_hline(yintercept = -1e4)+
  labs(x = NULL, y = 'Chromosome X (Mbp)', fill = 'SCNA') +
  guides(fill = guide_legend(nrow = 1)) +
  new_scale_fill() +
  geom_rect(aes(
    xmin = seq - 0.5,
    xmax = seq + 0.5,
    ymin = -5,
    ymax = 0,
    fill = frac2
  )) +
  scale_fill_viridis_c(direction = -1, limits = c(0, 1)) +
  scale_y_continuous(
    breaks = pretty_breaks(n = 7),
    limits = c(-5, 156040895 / 1e6),
    expand = c(0, 0)
  ) +
  scale_x_continuous(breaks = pretty_breaks(n = 7), expand = c(0, 0)) +
  theme_ipsum_rc(
    base_size = 13,
    axis_title_just = 'm',
    axis_title_size = 15,
    plot_margin = margin(5.5, 5.5, 5.5, 5.5),
    ticks = T,
    grid = F
  ) +
  theme(
    panel.spacing.x = unit(0.1, 'cm'),
    strip.text.x = element_text(hjust = 0.5, face = 'bold'),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    legend.position = "bottom",
    legend.direction = 'horizontal',
    legend.key.width = unit(2, 'cm')
  ) +
  panel_border(color = '#cccccc', size = 0.8) +
  facet_wrap(~Subtype, scales = 'free_x') +
  labs(fill = 'Subclonal Fraction\n')


# WGD status
p3 <- scna_chrx %>%
  left_join(
    ngspurity %>% select(Tumor_Barcode, MCN_WGD)
  ) %>%
  ggplot() +
  geom_rect(aes(
    xmin = seq - 0.5,
    xmax = seq + 0.5,
    ymin = -1,
    ymax = 1,
    fill = as.factor(MCN_WGD)
  )) +
  scale_fill_manual(values = ncicolpal[c(14, 18)]) +
  #geom_hline(yintercept = -1e4)+
  labs(x = NULL, y = 'Chromosome X (Mbp)', fill = 'SCNA') +
  guides(fill = guide_legend(nrow = 1)) +
  scale_y_continuous(
    breaks = pretty_breaks(n = 7),
    limits = c(-20, 20),
    expand = c(0, 0)
  ) +
  scale_x_continuous(breaks = pretty_breaks(n = 7), expand = c(0, 0)) +
  theme_ipsum_rc(
    base_size = 13,
    axis_title_just = 'm',
    axis_title_size = 15,
    plot_margin = margin(5.5, 5.5, 5.5, 5.5),
    ticks = T,
    grid = F
  ) +
  theme(
    panel.spacing.x = unit(0.1, 'cm'),
    strip.text.x = element_text(hjust = 0.5, face = 'bold'),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    legend.position = "bottom",
    legend.direction = 'horizontal'
  ) +
  #panel_border(color = '#cccccc',size=0.8)+
  facet_wrap(~Subtype, scales = 'free_x') +
  labs(fill = 'WGD status\n')

plot_grid(p1, p2, p3, align = 'v', axis = 'lr', nrow = 3)

```

##Fig. 2c
XIST mRNA expression across TCGA tumor types (male subjects), with TGCTs showing the highest levels.


```{r}

load('../data/tcga_xist_exp.RData',verbose = T)
mvalue <- xist_exp_new %>%
  group_by(Study) %>%
  summarise(mvalue = median(log2(RSEM + 1)))

tmp <- xist_exp_new %>%
  select(Study, Barcode) %>%
  unique() %>%
  count(Study) %>%
  mutate(StudyLab = paste0(Study, ' (', n, ')'))
p1 <- xist_exp_new %>%
  left_join(mvalue) %>%
  left_join(tmp) %>%
  mutate(StudyLab = fct_reorder(StudyLab, RSEM)) %>%
  ggplot(aes(StudyLab, RSEM, fill = mvalue)) +
  geom_boxplot(outlier.shape = NA, width = 0.7, size = 0.25) +
  scale_y_continuous(breaks = pretty_breaks(n = 7)) +
  scale_fill_material("deep-orange") +
  #facet_grid(.~Group,scales = 'free_x',space='free')+
  labs(
    y = 'XIST mRNA expression\nDESeq2 standardized RSEM (log2)',
    x = '',
    color = 'Group'
  ) +
  theme_ipsum_rc(
    axis_title_just = 'm',
    axis_title_size = 14,
    ticks = T,
    grid = 'Y'
  ) +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1),
    panel.spacing = unit(0.4, "lines"),
    strip.background = element_blank(),
    strip.text = element_text(hjust = 0.5),
    plot.margin = margin(4, 4, -6, 4)
  ) +
  guides(fill = "none") +
  panel_border(color = 'black', linetype = 1) +
  panel_border(color = 'black', size = 0.5) +
  theme(legend.position = 'none')

p1

# ggsave(
#   filename = 'tcga_xist_exp_tcga_gtex_p1.pdf',
#   plot = p1,
#   width = 6,
#   height = 5,
#   device = cairo_pdf
# )

```

##Fig. 2d
XIST mRNA expression in normal testis (GTEx), non-seminomas, and seminomas (TCGA). P-values from two-sided Wilcoxon rank-sum tests are shown above boxplots.


```{r, fig.width=4}
load('../data/tcga_xist_exp.RData',verbose = T)
my_comparisons <- list(
  c('Seminoma', 'Non-Seminoma'),
  c('Seminoma', 'GTEx Testis'),
  c('GTEx Testis', 'Non-Seminoma')
)

stat.test <- xist_exp2_testis %>%
  rstatix::wilcox_test(RSEM ~ Subtype, comparisons = my_comparisons) %>%
  rstatix::add_xy_position(x = "Subtype") %>%
  mutate(myformatted.p = sprintf("P = %.2e", p)) %>%
  mutate(SubtypeLab = NA)

stat.test
tmp <- xist_exp2_testis %>%
  select(Subtype, Barcode) %>%
  unique() %>%
  count(Subtype) %>%
  mutate(SubtypeLab = paste0(Subtype, ' (', n, ')'))

p2 <- xist_exp2_testis %>%
  left_join(tmp) %>%
  ggplot(aes(SubtypeLab, RSEM, fill = SubtypeLab)) +
  ggbeeswarm::geom_quasirandom(
    pch = 21,
    size = 2,
    width = 0.2,
    color = "white",
    stroke = 0.2
  ) +
  geom_boxplot(width = 0.7, outlier.shape = NA, alpha = 0.6, size = 0.4) +
  scale_fill_manual(
    values = as.character(c('GTEx Testis' = '#01665e', rev(subtypecol)))
  ) +
  scale_y_continuous(breaks = pretty_breaks(n = 7)) +
  labs(
    y = 'XIST mRNA expression\nDESeq2 standardized RSEM (log2)',
    x = '',
    color = 'Group'
  ) +
  theme_ipsum_rc(
    axis_title_just = 'm',
    axis_title_size = 14,
    ticks = T,
    grid = 'Y'
  ) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
    panel.spacing = unit(0.4, "lines"),
    strip.background = element_blank(),
    strip.text = element_text(hjust = 0.5),
    plot.margin = margin(4, 4, -6, 4)
  ) +
  guides(fill = "none") +
  panel_border(color = 'black', linetype = 1) +
  panel_border(color = 'black', size = 0.5) +
  theme(legend.position = 'none') +
  stat_pvalue_manual(
    stat.test,
    label = "myformatted.p",
    color = ncicolpal[1],
    size = 3.5
  )

p2
```

##Fig. 2e
Negative correlation between XIST expression and chromosome X median DNA methylation in seminomas (left) and non-seminomas (right). Pearson correlation coefficient (R) and P-value are shown below.


```{r,fig.width=10}

load('../data/tcga_xist_exp.RData',verbose = T)
load('../data/chrx_median_mtdata.RData',verbose = T)
# tmp <- tdata %>%
#   filter(CpG_chrm == 'chrX') %>% 
#   group_by(Name) %>%
#   summarize(MAD = mad(Beta, na.rm = TRUE)) %>%
#   arrange(desc(MAD)) %>% 
#   filter(MAD>0.1)
# mtdata <- tdata %>% filter(CpG_chrm == 'chrX',Name %in% tmp$Name) %>% group_by(Barcode) %>% summarise(mvalue=median(Beta,na.rm=T)) %>% ungroup()

#mtdata <- tdata %>% filter(CpG_chrm == 'chrX', Name=='cg17279685') %>% select(Barcode,mvalue=Beta)

pdata <- xist_exp_tgct %>% 
  left_join(mtdata %>% mutate(TCGA_Barcode = str_sub(Barcode,1,15)))

pdata %>% filter(!is.na(mvalue))%>% group_by(Subtype) %>% do(tidy(cor.test(log2(.$RSEM+1),.$mvalue)))

pdata %>% 
  ggplot(aes(log2(RSEM+1),mvalue))+
  geom_point()+
  facet_wrap(~Subtype,scales = 'free')+
  geom_smooth(method = 'lm')


pdata %>% 
  ggplot(aes(log2(RSEM+1),mvalue,fill=Subtype))+
  geom_point(pch=21,size=3,color="black",stroke=0.4)+
  geom_smooth(method = 'lm')+
  facet_wrap(~Subtype,scales = 'free')+
  scale_fill_manual(values = subtypecol)+
  scale_x_continuous(breaks = pretty_breaks(n = 7))+
  scale_y_continuous(breaks = pretty_breaks(n = 7))+
  labs(x='XIST mRNA expression log2(RNA Seq V2 RSEM + 1)',y='Chromsome X DNA methylation level (Beta)\n',color='Group')+
  theme_ipsum_rc(axis_title_just = 'm',axis_title_size = 14,ticks = T,grid = 'Y')+
  theme(panel.spacing = unit(0.4, "lines"),strip.background = element_blank(),strip.text = element_text(hjust = 0.5,face = 'bold'),plot.margin = margin(4,4,0,4))+
  guides(fill="none")+
  panel_border(color = 'black',linetype = 1)+  panel_border(color = 'black',size = 0.5)+
  theme(legend.position = 'none')+
  stat_cor(label.y = c(0.2,0.4),col = ncicolpal[1],size=4.5) 
#stat_regline_equation(label.y = c(1.8,3)) 

#ggsave(filename = 'XIST_chrX_methylation_level.pdf',width = 8,height = 4,device = cairo_pdf)
```

##Fig. 2f
Replication stress scores across TCGA tumor types based on replication stress gene signatures, with TGCTs showing the highest levels.


```{r}
load('../data/tcga_repstress_zscore.RData',verbose = T)

mvalue <- tcga_repstress_score_cbioportal %>% group_by(Study) %>% summarise(mvalue=median(RepStress_ZScore))
tmp <- tcga_repstress_score_cbioportal %>% select(Study,TCGA_Barcode) %>% unique() %>% count(Study) %>% mutate(StudyLab=paste0(Study,' (',n,')'))
tcga_repstress_score_cbioportal %>% 
  left_join(tmp) %>% 
  left_join(mvalue) %>% 
  mutate(StudyLab=fct_reorder(StudyLab,RepStress_ZScore)) %>% 
  ggplot(aes(StudyLab,RepStress_ZScore,fill=mvalue))+
  geom_boxplot(width=0.7,size=0.25,outlier.stroke = 0.1,outlier.size = 1,outlier.shape = 21,outlier.fill = NA)+
  scale_y_continuous(breaks = pretty_breaks(n=7),limits = c(-5,5))+
  scale_fill_material("deep-orange")+
  #facet_grid(.~Group,scales = 'free_x',space='free')+
  labs(y='Repstress Z-score',x='',color='Group')+
  theme_ipsum_rc(axis_title_just = 'm',axis_title_size = 14,ticks = T,grid = 'Y')+
  theme(axis.text.x = element_text(angle = 90,hjust = 1,vjust = 1),panel.spacing = unit(0.4, "lines"),strip.background = element_blank(),strip.text = element_text(hjust = 0.5),plot.margin = margin(4,4,-6,4))+
  guides(fill="none")+
  panel_border(color = 'black',linetype = 1)+  panel_border(color = 'black',size = 0.5)+
  theme(legend.position = 'none')


```

##Fig. 2g
Correlation between XIST expression and replication stress scores in seminomas (left) and non-seminomas (right), with significant positive correlation in seminomas only. Pearson R and P-value are shown above. 
All boxplots display median, IQR, and whiskers extending to 1.5Ã— IQR.


```{r,fig.width=10}
load('../data/tcga_repstress_zscore.RData',verbose = T)
tdata <- left_join(
  tcga_repstress_score_cbioportal %>% filter(Study == 'TGCT'),
  xist_exp_tgct %>%
    select(TCGA_Barcode, Subtype, RSEM) %>%
    mutate(RSEM = log2(RSEM + 1))
) %>%
  filter(!is.na(RSEM))


tdata %>%
  ggplot(aes(RSEM, RepStress_ZScore, fill = Subtype)) +
  geom_point(pch = 21, size = 3, color = "black", stroke = 0.4) +
  geom_smooth(method = 'lm') +
  facet_wrap(~Subtype, scales = 'free') +
  scale_fill_manual(values = subtypecol) +
  scale_x_continuous(breaks = pretty_breaks(n = 7)) +
  scale_y_continuous(breaks = pretty_breaks(n = 7)) +
  labs(
    x = 'XIST mRNA expression log2(RNA Seq V2 RSEM + 1)',
    x = '',
    color = 'Group'
  ) +
  theme_ipsum_rc(
    axis_title_just = 'm',
    axis_title_size = 14,
    ticks = T,
    grid = 'Y'
  ) +
  theme(
    panel.spacing = unit(0.4, "lines"),
    strip.background = element_blank(),
    strip.text = element_text(hjust = 0.5),
    plot.margin = margin(4, 4, 0, 4)
  ) +
  guides(fill = "none") +
  panel_border(color = 'black', linetype = 1) +
  panel_border(color = 'black', size = 0.5) +
  theme(legend.position = 'none') +
  stat_cor(label.y = c(2.3, 4.2))

```

## R session information
```{r}
sessionInfo()
```


